assets/js/theme/custom/update-product-image-based-on-options.js

```diff
@@ -0,0 +1,95 @@
+/**
+ * Updates the main product image based on selected product options using GraphQL.
+ */
+export default function updateProductImageBasedOnOptions(storefrontApiToken, productId) {
+    const mainProductImage = document.querySelector('.js-main-image-link')
+    if (!mainProductImage) {
+        return
+    }
+
+    const productOptions = document.querySelector('.js-product-option-change')
+    if (!productOptions) {
+        return
+    }
+
+    function getOptionId(id) {
+        const matches = id.match(/\[(\d+)\]/)
+        return matches ? matches[1] : null
+    }
+
+    function getSelectedOptions() {
+        const selectedOptions = {}
+        productOptions.querySelectorAll('select, input[type="radio"]:checked').forEach((input) => {
+            const optionId = getOptionId(input.name)
+            if (optionId) {
+                selectedOptions[optionId] = input.value
+            }
+        })
+        return selectedOptions
+    }
+
+    function applyImageUpdate(variants) {
+        const selectedOptions = getSelectedOptions()
```

Comment:
Too much happens in this function. Let's create a separate function called `normalizeVariantsOptions` that will take a data and return some nice object . Also, correct me if I'm wrong, but... I think each option's input has its own unique ID, so you may store just an array of selected values (`[118,116]`) instead of having more complex object with option group ID as a key? 

So if you could store just ids of selected inputs, the whole comparing process gets so much easier, because then you can have:

```
const variants = {
  [variantID] : {
     defaultImage: [default image url]
      options: [ single_option_ID, ... ]
    }, 
  (...)
}
```

and simply find a match by ```variants.find((variant) => selectedOptions.every((option) => variant.options.includes(option)))```

Much simpler and cleaner :)